### PR TITLE
Reworks & TGUIs mecha equipment

### LINKED ISSUE #39918: Gygax leg actuators function on relay power
[Round ID]: # 93013

[Reproduction]: # Build gygax, put relay in it, fulfill the dream murderboning with infinite leg actuator power.

### LINKED ISSUE #40782: H.O.N.K. inteface has double text
image

### PR DESCRIPTION
**Pull request by TiviPlus**
### TLDR
Mech UI now TGUI: https://streamable.com/ahjydy
Mechs now use "slots" for equipment
They can only mount 2 guns(left and right arm, left and right click to use)
Tesla & Concealed bay is removed
Removed mech damage deflection

### Full list of changes
Note: weapons in this section refer to all click-using equipment such as guns/hydraulic clamp/etc

Hackmd: https://hackmd.io/lgr-LetfSKyHzPP0zQUXPA
Tesla has been removed (Tesla gives you effectively infinite power on station and off station(where the design direction for mechs is intended) they are useless)
Concealed mech bay has been removed (Concealed mecha bay relied on selection for visibility so it doent make sense to exist when all equips are "selected", the code was bad (instead of flag or something you just used locate() everywhere)
Honker no longer has snowflake UI (I'm sorry but if I have to make a third UI for this im going to go insane)
Mechs now mount 2 arm weapons, and certain amount of "utility", "armor" and "power" module types
The two arm weapons do not share a cooldown anymore UNLESS they are identical weapons
left click to shoot left weapon right click for right weapon, this will shoot the weapon unless you click next to you with a ranged weapon it will then try to melee (This also applies to attaching the weapon to the mech)
It is no longer possible to deflect or have multipliers for damage from innate or armor sources, armor will now just apply a direct armor change
Since plasma generator needed to be selected to use you now need to click on the power generator with a stack of plasma to refill it
Internal damage:
The mech no longer needs to be low health to take internal damage
Internal damage now no longer has several convoluted ways to fix each different type and is now a timed action performed in the UI
Weapons no longer have an RNG chance to be damaged or immediately deleted by damage
Weapons will now take damage aimed at the right and left arms respectively, this can be repaired using the new UI. NOTE: this is intended to be implemented as weapon/module disabling but I left it out of this PR to try resemble some shortness
Both internal and equipment damage have minimum thresholds of damage required in a hit before they attempt to check whether the mech should take internal or equipment damage, teh threshold is lower for equipment damage
You can no longer reload weapons using energy

### Demo Video
https://streamable.com/ahjydy

## Why It's Good For The Game
better mech UI
better mech controls
more intuitive
changing of OP/do-it-all loadouts encourages more balanced mech loadouts
Down the road better mech controls are a prerequisite if mechs are intended to be the primary method for lavaland mechs that are invisioned, since clunkyness would make the job instantly less fun, and mech balance being wack would be annoying too

Testmerge this pr pls
I can kill more mech shitcode after this is merged (though notably missing here my PR for cleaning equipment code in prep for this got staleclosed because noone reviewed lol)
The armor sprites are not the best because I need this PR first to expand what I want to do with armor while my spriter finishes the sprites

## Changelog
Closes https://github.com/tgstation/tgstation/issues/40782
Closes https://github.com/tgstation/tgstation/issues/39918

:cl:
refactor: The mech UI(s) now use TGUI
add: Mechs no longer have RNG chances to completely ignore attacks
add: Mech armor will now directly add armor instead increasing deflection chance
add: Mechs now have slots for equipping "power", "utility" and "armor" modules
add: Mechs now use left and right click to use the left and right weapons respectively
add: Mech internal damage no longer requires low HP and instead now requires a minimum damage in an attack to consider it
del: hidden mech bay has (at least for now) been removed
del: Some ballistic weapons reloading with energy has been removed so they have infinite ammo
del: tesla power module has been removed
del: Honker no longer has snowflake UI that plays sounds
fix: admin uis now work on mech equipment properly
/:cl:

### CODE REVIEW

**Comment by TheBonded** [comment]
why do they still share a cooldown if they are the same weapon? having an lmg and a shotgun lets you shoot more than having two lmgs?

**Comment by moocowswag** [comment]
> > why do they still share a cooldown if they are the same weapon? having an lmg and a shotgun lets you shoot more than having two lmgs?
> 
> The objective is to prevent a similar situation to before the shared cooldown where you would take the same equipment 3 times then rapidly switch through to min max one weapon

What about a heat system where each shot generates heat depending on the weapon that was fired and you arnt allowed to overheat? I know its a lot to ask for but it could be a way to limit someone from just using 2 strong high dps identical weapons without having some sort of cooling equipment.

**Comment by moocowswag** [comment]
Also I think one of the nukie mechs uses the tesla energy relay.

**Comment by SmArtKar** [comment]
there's zero actual reason for concealed mech bay to go, it's a nice robo-unique item and it doesn't interfere with your arms and utility slot system, just make it into a utility item or make it attachable to a gun module to make it installable into normal mechs

File: code/modules/vehicles/mecha/_mecha.dm [hunk 1]
```diff
@@ -750,23 +788,53 @@
...
+		if(MECHA_INT_TEMP_CONTROL)
+			return "Resetting temperature module..."
+		if(MECHA_INT_TANK_BREACH)
+			return "Activating tank sealant..."
+		if(MECHA_INT_CONTROL_LOST)
+			return "Recalibrating coordination system..."
+
+///gets the on-fail balloon alert flufftext
+/obj/vehicle/sealed/mecha/proc/get_int_repair_fluff_end(flag)
+	switch(flag)
```
Comment (Sealed101) [inline code/modules/vehicles/mecha/_mecha.dm:817]: messed the comments with this proc and the next one

File: code/modules/vehicles/mecha/equipment/tools/medical_tools.dm [hunk 2]
```diff
@@ -309,15 +307,31 @@
...
+	)
+
+/obj/item/mecha_parts/mecha_equipment/medical/syringe_gun/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+	if(action == "change_mode")
+		mode = !mode
+		return TRUE
+	else if(action == "show_reagemts")
```
Comment (Sealed101) [inline code/modules/vehicles/mecha/equipment/tools/medical_tools.dm:332]: reagemts

File: tgui/packages/tgui/interfaces/Mecha/ArmPane.tsx [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+              ref: props.weapon.ref,
+              gear_action: "change_mode",
+            })} />
+        </LabeledList.Item>
+        <LabeledList.Item label={"Reagent control"}>
+          <Button
+            content={"View"}
+            onClick={() => act('equip_act', {
+              ref: props.weapon.ref,
+              gear_action: "show_reagemts",
```
Comment (Sealed101) [inline tgui/packages/tgui/interfaces/Mecha/ArmPane.tsx:217]: ditto

**Comment by Sealed101** [comment]
from what i can see, you can equip even more armor modules now, which sounds badass and probably broken in some cases

mecha damage zones seem cool but without some way to telegraph it to the players it will be obscure, unless i missed it

tgui is a blessing but the strength of the ui we have now is tabs and keeping the most important data open at all times. while this ui has the same data, it opens up a whole lot of options which might not be required to be monitored at all times. also html ui was cool to leave hanging over the statpanel, and i can't quite see this ui fit in that place.

also have to ask but does this fix the mecha id assignment properly i.e. do posibrains or mmis piloting the mech gain access when its provided to the mech? iirc its still a problem

**Comment by thgvr** [comment]
are AIs able to use mech equipment with this?

**Comment by Ghommie** [comment]
The HONK interface double text bug was actually fixed by me roughly a year ago by the by, I simply didn't notice the issue report (which should be closed already).

**Comment by CourageTheCat** [comment]
FINALLY I can manage my mech mid combat without byond freezing and dying

File: code/modules/vehicles/mecha/equipment/mecha_equipment.dm [hunk 4]
```diff
@@ -102,52 +87,81 @@
...
 	var/C = chassis.loc
 	. = do_after(user, delay, target=target)
-	if(!chassis || chassis.loc != C || src != chassis.selected || !(get_dir(chassis, target)&chassis.dir))
+	if(!chassis || chassis.loc != C || !(get_dir(chassis, target)&chassis.dir))
 		return FALSE
 
-/obj/item/mecha_parts/mecha_equipment/proc/can_attach(obj/vehicle/sealed/mecha/M)
-	if(LAZYLEN(M.equipment)<M.max_equip)
+/obj/item/mecha_parts/mecha_equipment/proc/can_attach(obj/vehicle/sealed/mecha/M, rightclickattach = FALSE)
+	return default_can_attach(M, rightclickattach)
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/equipment/mecha_equipment.dm:102]: snake_case for `rightclickattach` please. 
Also it might be clearer to rename to something like "attach_to_right" or something. `rightclickattach` isn't a very clear var name.

File: code/modules/vehicles/mecha/combat/honker.dm [hunk 5]
```diff
@@ -5,140 +5,24 @@
...
-
 /obj/vehicle/sealed/mecha/combat/honker/play_stepsound()
 	if(squeak)
 		playsound(src, "clownstep", 70, 1)
 	squeak = !squeak
 
-/obj/vehicle/sealed/mecha/combat/honker/Topic(href, href_list)
-	..()
-	if (href_list["play_sound"])
-		switch(href_list["play_sound"])
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/combat/honker.dm:114]: this removal needs to be in the changelog

File: code/modules/vehicles/mecha/equipment/mecha_equipment.dm [hunk 6]
```diff
@@ -7,6 +7,7 @@
 	icon_state = "mecha_equip"
 	force = 5
 	max_integrity = 300
+	var/equipment_slot = MECHA_WEAPON
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/equipment/mecha_equipment.dm:10]: Auto-doc new vars please.

**Review by MrMelbert — COMMENTED** [review]
I like this

**Comment by castawaynont** [comment]
make the tab have a vertical option

**Comment by Serxule** [comment]
the tgui is a bit of a menu vomit right now, but looks better, but can we have more details on equipment slots? like do all mechs have 2 armor slots/gun slots etc? and also how will the new armor work now? (also rng deflection is a thing with mechs regardless of armor), but what is it changed to now? just damage reduction? and you said power will be made meaningful? will t4 parts still not drain power? also what does that even mean? you really need to explain the changes better, because i don't really know what you're doing

**Comment by Serxule** [comment]
> I think one of the nukie mechs uses the tesla energy relay (which I think shouldn`t be removed, it should probably absorb a lot more power with a really bad efficiency so every 5kw from an apc gives you 1kw in your mech for example, basically acting like a crappy power sink)

also, nukie mech equipment is pretty shit, the equipment needs to be replaced with actually good equipment

**Comment by githubuser4141** [comment]
Is it possible to resize the TGUI mech menu? It's pretty important to have the mech menu open when you're piloting a mech to properly keep track of your equipment/power/health, it'd be annoying to just have to guess.

File: code/modules/vehicles/mecha/equipment/mecha_equipment.dm (hunk 6 above)
Comment (TiviPlus) [reply code/modules/vehicles/mecha/equipment/mecha_equipment.dm:11]: Done

File: code/modules/vehicles/mecha/combat/honker.dm (hunk 5 above)
Comment (TiviPlus) [reply code/modules/vehicles/mecha/combat/honker.dm:114]: Done

File: code/modules/vehicles/mecha/equipment/mecha_equipment.dm (hunk 4 above)
Comment (TiviPlus) [reply code/modules/vehicles/mecha/equipment/mecha_equipment.dm]: Done

File: tgui/packages/tgui/interfaces/Mecha/ArmPane.tsx (hunk 3 above)
Comment (TiviPlus) [reply tgui/packages/tgui/interfaces/Mecha/ArmPane.tsx]: Done

File: code/modules/vehicles/mecha/equipment/tools/medical_tools.dm (hunk 2 above)
Comment (TiviPlus) [reply code/modules/vehicles/mecha/equipment/tools/medical_tools.dm]: Done

File: code/modules/vehicles/mecha/_mecha.dm (hunk 1 above)
Comment (TiviPlus) [reply code/modules/vehicles/mecha/_mecha.dm]: Done

**Comment by TiviPlus** [comment]
> What about a heat system where each shot generates heat depending on the weapon that was fired and you arnt allowed to overheat?
 
Wouldnt fix the issue but Ill jot it down as a potential weapon idea

>basically acting like a crappy power sink

The vast majority of teslas are produced by roboticists for their own mechs, that wouldn't make much sense

>you can equip even more armor modules now, which sounds badass and probably broken in some cases

before: 
Assumptions: savannah with 1 ranged armor upgrade, laser gun attacking from front
43.125% chance for 0 damage
32.66% damage taken on a hit

After:
Assumptions: savannah with 3 ranged armor upgrade, laser gun attacking from front
30% average damage on hit

>mecha damage zones seem cool but without some way to telegraph it

This PR only adds the barebones required TGUI code for repair and so
I'm going to work on damage overlays and a UI damage doll as a seperate PR after this

>double text bug was actually fixed by me roughly a year ago 

Ye theres a lot of lost bugs that were fixed I just added it here cus its a UI bug 

>Is it possible to resize the TGUI mech menu?

Yes, and health/power/internal damage is also an ingame HUD icon, as mentioned above im working on a doll for mechs

>how will the new armor work now?
>what is it changed to now
>will t4 parts still not drain power

Please read the PR the changes are listed right there

File: code/modules/vehicles/mecha/equipment/weapons/weapons.dm [hunk 7]
```diff
@@ -464,40 +445,34 @@
...
+	. = ..()
+	if(action == "toggle")
 		harmful = !harmful
-		if(!chassis)
-			return
 		if(harmful)
 			to_chat(usr, "[icon2html(src, usr)][span_warning("Lethal Fisting Enabled.")]")
 		else
-			to_chat(usr, "[icon2html(src, usr)][span_warning("Lethal Fisting Disabled.")]")
+			to_chat(usr, "[icon2html(src, usr)][span_warning("Patting Disabled.")]")
```
Comment (Fikou) [inline code/modules/vehicles/mecha/equipment/weapons/weapons.dm:474]: patting?

Comment (TiviPlus) [reply code/modules/vehicles/mecha/equipment/weapons/weapons.dm:474]: fixed

**Comment by githubuser4141** [comment]
I didn't know the old boosters increased deflect that much, I thought it was only about 1.15x the original deflect? Anyways, I guess this is more about mechanics rather then balance, but removing deflects is pretty harsh, especially on the Phazon mech which previously had a solid 30% chance without deflectors. Now it's just a Gygax but with less health, slower max speed, and it's punches do less damage and don't have knockdown. Sure, it can phase, but that's costly and other mechs can just punch through stuff.

**Comment by Serxule** [comment]
> Please read the PR the changes are listed right there

missed that since its a masssive wall of text, and i see nothing about power or parts, also being able to target sections of a mech to (im assuming easily) break them seems fairly unbalanced, and a big nerf to mechs

File: tgui/packages/tgui/interfaces/Mecha/MechStatPane.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+              </Button>
+            </LabeledList.Item>
+            <LabeledList.Item label="Cabin pressure">
+              <Box color={cabin_pressure>cabin_dangerous_highpressure?"red":null}>
+                {cabin_pressure} kPa
+              </Box>
+            </LabeledList.Item>
+            <LabeledList.Item label="Cabin temperature">
+              <Box>
+                {GetTempFormat(cabin_temp)}
```
Comment (spookydonut) [inline tgui/packages/tgui/interfaces/Mecha/MechStatPane.tsx:63]: you might need to double check with @stylemistake that it's ok to use a function like this inside the template.

File: code/modules/vehicles/mecha/working/ripley.dm [hunk 9]
```diff
@@ -189,16 +181,37 @@
...
+	return data
+
+/obj/item/mecha_parts/mecha_equipment/ejector/ui_act(action, list/params)
+	. = ..()
+	if(action == "eject")
+		var/obj/vehicle/sealed/mecha/working/ripley/miner
+		var/obj/crate = locate(params["cargoref"]) in miner.cargo
+		if(!crate)
+			return FALSE
+		to_chat(miner.occupants, "[icon2html(src,  miner.occupants)][span_notice("You unload [crate].")]")
```
Comment (spookydonut) [inline code/modules/vehicles/mecha/working/ripley.dm:207]: miner is unset here

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/Mecha/MechStatPane.tsx (hunk 8 above)
Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/Mecha/MechStatPane.tsx:63]: Surely `GetTempFormat` is a pure function? Pure functions in render are fine.

**Review by spookydonut — APPROVED** [review]

File: code/modules/vehicles/mecha/_mecha.dm [hunk 10]
```diff
@@ -361,12 +396,9 @@
...
 		else
 			. += "It's falling apart."
-	var/hide_weapon = locate(/obj/item/mecha_parts/concealed_weapon_bay) in contents
-	var/hidden_weapon = hide_weapon ? (locate(/obj/item/mecha_parts/mecha_equipment/weapon) in equipment) : null
-	var/list/visible_equipment = equipment - hidden_weapon
-	if(length(visible_equipment))
+	if(LAZYLEN(flat_equipment))
 		. += "It's equipped with:"
-		for(var/obj/item/mecha_parts/mecha_equipment/ME in visible_equipment)
+		for(var/obj/item/mecha_parts/mecha_equipment/ME in flat_equipment)
```
Comment (Watermelon914) [inline code/modules/vehicles/mecha/_mecha.dm:401]:
```suggestion
		for(var/obj/item/mecha_parts/mecha_equipment/ME as anything in flat_equipment)
```

File: code/modules/vehicles/mecha/equipment/tools/medical_tools.dm [hunk 11]
```diff
@@ -60,14 +84,13 @@
 	chassis.visible_message(span_warning("[chassis] starts putting [target] into \the [src]."))
 	if(!do_after(source, equip_cooldown, target=target))
 		return
-	if(!chassis || src != chassis.selected || !(get_dir(chassis, target)&chassis.dir))
+	if(!chassis || !(get_dir(chassis, target)&chassis.dir))
```
Comment (Watermelon914) [inline code/modules/vehicles/mecha/equipment/tools/medical_tools.dm:87]:
```suggestion
	if(!chassis || !(get_dir(chassis, target) & chassis.dir))
```

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

**Comment by TheSilverNuke** [comment]
>add: Mechs now use left and right click to use the left and right weapons respectively

So what happens when an AI is inserted into the mech via control beacon, mech override, or card? It's bound to MMB currently because they still need to click on doors and other intractable items.

File: code/modules/vehicles/mecha/_mecha.dm [hunk 12]
```diff
@@ -1204,7 +1263,7 @@
 		return FALSE
 	var/ammo_needed
 	var/found_gun
-	for(var/obj/item/mecha_parts/mecha_equipment/weapon/ballistic/gun in equipment)
+	for(var/obj/item/mecha_parts/mecha_equipment/weapon/ballistic/gun in flat_equipment)
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/modules/vehicles/mecha/_mecha.dm:1266]: Typecheck required

**Comment by InvalidArgument3** [comment]
The hackmd sounds incredible, having hotkeys would be amazing
>They can only mount 2 guns(left and right arm, left and right click to use)
>locational targeting
I'm so fucking hyped to play mechwarrior 2 in spess

File: code/modules/vehicles/mecha/_mecha.dm (hunk 12 above)
[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/vehicles/mecha/_mecha.dm:1266]: yep I see

File: code/modules/vehicles/mecha/equipment/mecha_equipment.dm [hunk 13]
```diff
@@ -29,58 +31,43 @@
...
+			detach(get_turf(src))
+			return TRUE
+		if("toggle")
+			activated = !activated
+			return TRUE
+		if("repair")
+			ui.close() // allow watching for baddies and the ingame effects
+			chassis.balloon_alert(usr, "starting repair")
+			while(do_after(usr, 1 SECONDS, chassis) && get_integrity() < max_integrity)
+				repair_damage(30)
```
Comment (Watermelon914) [inline code/modules/vehicles/mecha/equipment/mecha_equipment.dm:67]: damn.. can't you do this in a different way like process()?
This seems like an easy way of doing it

Comment (Watermelon914) [reply code/modules/vehicles/mecha/equipment/mecha_equipment.dm:67]: though that's a bit of an overexaggeration, it's not that bad but I feel like you could do this without a while() sleep() loop

File: code/modules/vehicles/mecha/working/ripley.dm [hunk 14]
```diff
@@ -189,16 +181,36 @@
...
+			"name" = crate.name,
+			"ref" = REF(crate),
+		))
+	return data
+
+/obj/item/mecha_parts/mecha_equipment/ejector/ui_act(action, list/params)
+	. = ..()
+	if(action == "eject")
+		var/obj/vehicle/sealed/mecha/working/ripley/miner = chassis
+		var/obj/crate = locate(params["cargoref"]) in miner.cargo
```
Comment (Watermelon914) [inline code/modules/vehicles/mecha/working/ripley.dm:203]:
```suggestion
	. = ..()
	if(.)
		return TRUE
	if(action == "eject")
		var/obj/vehicle/sealed/mecha/working/ripley/miner = chassis
		var/obj/crate = locate(params["cargoref"]) in miner.cargo
```

**Comment by SpookyTheFox** [comment]
I'm investigating a possible bug related to this PR. It seems I could not card an AI out of a mech using the methods before, which was to enable maintenance protocol then swipe the intelicard on the mech. Has that procedure changed, or is there any other feature discontinuity there?

**Comment by TheSilverNuke** [comment]
Literally as expected AI intellicards are unable to use their mech modules. Good job. I brought this up while the PR is open but it seems this simple oversight was too much for you to fix.
